File: Python/k08_2sub.py
from Golem import Golem

def attack_golem(lv, playerHp):
    golem = Golem(lv * 50 + 100, lv * 10 + 40, lv * 10 + 30)
    _playerHp = playerHp
    attack_method = {
        'Attack': 65,
        'Skill': 99,
        'Magic': 133
        }

    print('ゴーレムLv.{}が現れた！'.format(lv + 1))
    while golem.hp > 0:
        print('\n残りHP：{}'.format(golem.hp))
        print('武器を選択してください（１．攻撃　２．特技　３．魔法）＞', end='')
        choice = int(input())
        damage_dispatch = {
            1: attack_method['Attack'],
            2: attack_method['Skill'],
            3: attack_method['Magic']
            }
        damage = damage_dispatch[choice] - golem.defense

        if damage <= 0:
            damage = 0
            print('ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」')
            print('ゴーレムがあなたを攻撃しました！攻撃値：{}'.format(golem.attack))
            _playerHp = damage_player(_playerHp, golem.attack)
            
            if _playerHp <= 0:
                _playerHp = 0
                print('あなたはゴーレムに負けました！ゲームオーバー！')
                return _playerHp
            print('あなたの残りHPは：{}'.format(_playerHp))

        print('ダメージは{}です。'.format(damage))
        golem.hp -= damage

        if golem.hp <= 0:
            golem.hp = 0

        if golem.hp == 0:
            print('ゴーレムLv.{}を倒した！'.format(lv + 1))
    return _playerHp

def damage_player(playerHp, attackPower):
    return playerHp - attackPower

File: Nodejs/K08_2sub.ts
import { Golem } from './Golem';
import * as readline from 'readline';
import { format } from 'util';

export function AttackGolem(lv: number, playerHp: number, rl?: readline.ReadLine): Promise<number> {
    try {
        var golem = new Golem(lv * 50 + 100, lv * 10 + 40, lv * 10 + 30);
        console.log(format("ゴーレムLv.%dが現れた！", lv + 1));
        let choice = 0;

        let selectAttackMethod = () => {
            return new Promise((resolve, reject) => {
                rl.question("武器を選択してください（１．攻撃　２．特技　３．魔法）＞", (value) => {
                    choice = Number.parseInt(value);
                    resolve();
                });
            });
        };

        return new Promise<number>(async (resolve, reject) => {
            while (golem.hp > 0) {
                console.log(format("\n残りHP：%d", golem.hp));
                await selectAttackMethod();
                let damage = 0;
                switch (choice) {
                    case 1:
                        damage = Number(AttackMethod.Attack);
                        break;
                    case 2:
                        damage = Number(AttackMethod.Skill);
                        break;
                    case 3:
                        damage = Number(AttackMethod.Magic);
                        break;
                    default:
                        break;
                }
                damage -= golem.defense;

                if (damage <= 0) {
                    damage = 0;
                    console.log("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
                    console.log(format("ゴーレムがあなたを攻撃しました！攻撃値：%d", golem.attack));
                    playerHp = DamagePlayer(playerHp, golem.attack);
                    if (playerHp <= 0) {
                        playerHp = 0;
                        console.log("あなたはゴーレムに負けました！ゲームオーバー！");
                        resolve(playerHp);
                        return;
                    }
                    console.log(format("あなたの残りHPは：%d", playerHp));
                }
                console.log(format("ダメージは%dです。", damage));
                golem.hp -= damage;

                if (golem.hp <= 0)
                    golem.hp = 0;
                if (golem.hp == 0)
                    console.log(format("ゴーレムLv.%dを倒した！", lv + 1));
            }
            resolve(playerHp);
        });

    } catch (e) {
        console.error(e);
    }
}

enum AttackMethod {
    Attack = 65,
    Skill = 99,
    Magic = 133
}

function DamagePlayer(playerHp: number, attackPower: number): number {
    return playerHp - attackPower;
}